head_test.go

```diff
@@ -72,3 +79,410 @@ func readPrometheusLabels(fn string, n int) ([]labels.Labels, error) {
 	}
 	return mets, nil
 }
+
+func TestHeadBlock_e2e(t *testing.T) {
+	dir, err := ioutil.TempDir("", "test_persistence_e2e")
+	require.NoError(t, err)
+	defer os.RemoveAll(dir)
+
+	lbls, err := readPrometheusLabels("testdata/20k.series", 20000)
+	require.NoError(t, err)
+
+	l := log.NewLogfmtLogger(os.Stdout)
+	ts := time.Now().UnixNano() / 1000000
+
+	hb, err := createHeadBlock(dir, 0, l, ts, ts+50*1000)
+	require.NoError(t, err)
+	mhb := newMockHead(dir)
+
+	// 10 scrapes of 10 targets.
+	for i := 0; i < 10; i++ {
+		for j := 0; j < 10; j++ {
+			ts := time.Now().UnixNano() / 1000000
+			go ingestScrapes(t, lbls, strconv.Itoa(j), ts, hb.Appender(), mhb.Appender())
+		}
+		time.Sleep(2 * time.Second)
+	}
+
+	require.Equal(t, mhb.meta.Stats.NumSamples, hb.meta.Stats.NumSamples,
+		"NumSamples: exp: %d, got: %d", mhb.meta.Stats.NumSamples, hb.meta.Stats.NumSamples)
+	require.Equal(t, mhb.meta.Stats.NumSeries, hb.meta.Stats.NumSeries, "NumSeries")
+	require.Equal(t, mhb.meta.Stats.NumChunks, hb.meta.Stats.NumChunks, "NumChunks")
+}
+
+func ingestScrapes(
+	t *testing.T,
+	l []labels.Labels,
+	instance string,
+	ts int64,
+	app Appender,
+	mockApp Appender,
+) {
+	for _, lset := range l {
+		nlset := make(labels.Labels, len(lset))
+		copy(nlset, lset)
+
+		nlset = append(nlset, labels.Label{Name: "instance", Value: instance})
+		sort.Sort(nlset)
+
+		val := 100 * rand.Float64()
+		_, err := app.Add(nlset, ts, val)
+		if err != nil {
+			app.Rollback()
+		}
+		require.NoError(t, err)
+
+		_, err = mockApp.Add(nlset, ts, val)
+		require.NoError(t, err)
+	}
+	require.NoError(t, app.Commit())
+	require.NoError(t, mockApp.Commit())
+	return
+}
+
+type mockSeries struct {
+	lset    labels.Labels
+	samples []sample
+}
+
+func (m mockSeries) Labels() labels.Labels { return m.lset }
+
+func (m mockSeries) Iterator() SeriesIterator {
+	return newListSeriesIterator(m.samples)
+}
+
+type mockHead struct {
+	dir  string
+	meta BlockMeta
+
+	postings *memPostings
+	values   map[string]stringset
+
+	// The series for this head. The refs are stored in index.
+	series []mockSeries
+	lref   lmap
+
+	activeWriters uint64
+
+	sync.RWMutex
+}
```

Comment:
I completely agree. So the whole point of this was to actually get the querier working. But it became quite bulky and I couldn't get to querier yet.

The primary motivation to implement a mockHead was for me to understand the behaviour better. By trying to implement a naaive version and then looking at the implementation to figure out the optimisations and details. It helped me understand the codebase really well.

If you think this is unnecessary then it is completely fine to not merge this in. And a huge thanks for reviewing.